### ISSUE TITLE
FsSrGen obsolete?

### ISSUE THREAD

**Issue by enricosada** [Microsoft/visualfsharp#794]
FsSrGen is used to generate an resx resource file from a txt file like

```
invalidOperationOnProvidedType,"The operation '%s' on item '%s' should not be called on provided type, member or parameter"
```

to a function `invalidOperationOnProvidedType : string -> string -> string` 

```fsharp
    /// The operation '%s' on item '%s' should not be called on provided type, member or parameter
    /// (Originally from D:\github\fsharp_coreclr\src\fsharp\FSharp.Data.TypeProviders\FSData.txt:13)
    static member invalidOperationOnProvidedType(a0 : System.String, a1 : System.String) = (GetStringFunc("invalidOperationOnProvidedType",",,,%s,,,%s,,,") a0 a1)
```

an example [/src/fsharp/FSharp.Data.TypeProviders/FSData.txt](
https://github.com/Microsoft/visualfsharp/blob/6bff0d341bc442b8b5dbc4cfca9683cc123d3737/src/fsharp/FSharp.Data.TypeProviders/FSData.txt) is compiled into `src\fsharp\FSharp.Data.TypeProviders\obj\net40\FSData.fs` and `src\fsharp\FSharp.Data.TypeProviders\obj\net40\FSData.resx`

actually the source for fssrgen is not in this repo and is in `lkg` as a binary, not a nuget reference.

FsSrGen need also [FSharp.SRGen.Build.Tasks.dll]() and `FSharp.SRGen.targets` to be used inside msbuild files.

Can we simplify the codebase? The change is low risk, we need only to commit inside the repo the generated files, and include these inside the projects instead of `FsSrGen Include="FSData.txt"`

Cons:
- need to maintain resx and .fs, but is already like that for fscore.dll

Pro:

- no maintenance of fssrgen tools ( if we need to change it? for example for coreclr? now is a binary )
- one less dll who depends on [lkg/FSharp-14.0.23413.0/bin/FSharp.Core](https://github.com/Microsoft/visualfsharp/tree/6bff0d341bc442b8b5dbc4cfca9683cc123d3737/lkg/FSharp-14.0.23413.0/bin/FSharp.Core) v 2.0.50726.900
- no need to run fssrgen on build, just include the files

**Comment by dsyme** [comment]
I've always thought this tool was a bit of a mistake, especially the FSharp.SRGen.Build.Tasks.dll.  I'd love to remove both, though admittedly it has made adding new error messages really very easy.

I'm not quite sure if there's an interaction with the localization process for F# in Visual Studio - which is done internally to Microsoft. VS 2010-2015 have been localized and my impression is that the localization into Japanese in particular has been important. But they probably did the localization by adding new .resx files.

**Comment by enricosada** [comment]
the idea is really nice, but maybe should be implemented with type provider not msbuild custom tasks ( i dont think these are supported by dnx ).

We can skip the .txt and use a type provider to read directly the resx and translate the arguments ( add xml comment ) to a typed function.

```xml
  <data name="invalidOperationOnProvidedType" xml:space="preserve">
    <!-- {0}: %s -->
    <value>The operation '{0}' on item '{1}' should not be called on provided type, member or parameter</value>
  </data>
```

but for now i think is easier to cleanup

**Comment by KevinRansom** [comment]
It can stay as it is for now.  It works, is not a problem and is a well understood workflow.  It is at best a priority 3 change.

The areas for F# that need most work is the IDE, and the test harness.  We still have too many different style of tests, different test harnesses and a massive test conceptual overhead, inability to debug tests, select individual test cases to run.

IDE improvements make the product more attractive, usable and will encourage new F# users more reason to stick with the language.

Just changing the way we do stuff that already works, is a cost.  Lets make new stuff or fix broken stuff.

**Comment by enricosada** [comment]
@KevinRansom This issue is about coreclr too, not maintenance only, or clean the sake of cleanup.

For coreclr ( full build on coreclr ) we need to rebuild fssrgen using coreclr fsharpcore (and others tools like `findunused.exe`, `fssrgen build task` ).
I opened this issue because i think is a good idea this time to remove it, if ms internally use resx as @dsyme said, instead of upgrade it. That the core of this issue.

FsLex,FsYacc is another problem (no source here), but there is a good maintained fork https://github.com/fsprojects/FsLexYacc so maybe we can use that, but the problem exists. If we are going to use https://github.com/fsprojects/FsLexYacc we can start now, use the nuget, and after fscore coreclr is release we need only to increment nuget version.

I know this is not high priority, but is a simple task, and for sure i am not asking you @KevinRansom to do it, or @otawfik-ms or @NumberByColors 

About roadmap, sry for rant, but ok, no need to rewrite it in each minor or trivial issue/pr, i know there are more priority tasks. Add a minor/trivial/prio-3 label, but for sure i am not asking you @KevinRansom to do it. It's a task a contributor can do. Sorry if you need to review it and merge it, but if we can in parallel do more work, the better i think.

About priorities ( this is not a rant, only a bit sad ):
- tests: before coreclr test suites were not a problems, some from community wrote that was a problem, to improve it, but was not a priority. If we improve the code ( not you, the community ), it's going to speed up development and help new contributors, we should try to understand weakness and discuss how to improve.
- ide: it's really difficult to contribute, not because is complicated ( see VFPT, lots of contributions, same requirements as extension ) but because is complicated for a contributor to start. Try the experience in roslyn or PTVS or nodejstools, is easier to download -> run. Other problems: 
  - no docs, visualfsharp wiki is private, so we community cannot improve it ( i tried )
  - old code ( c#, vb, f# mixed but i have seen worse at work, lot worse ) 
  - roslyn/vs is going another way ( project.json, xproj, CPS, our plan about that? ). 
My personal experience, i tried three times (pr) to fix single a bug, all rejected because of fear of breaking. Small fix doesnt help with big problems.

Work in progress:
- tests: there are 2 pr waiting, one for coreclr unit tests, one for cambridge suite, after these are merged i'll continue with coreclr cambridge and nunit fsharpqa
- ide: i have a pr waiting for @mousetraps / @zooba, but i think i can update the pr, useless to wait more, i'll update and we can start merge PTVS code. I fear is going to be useless work because of the new CPS, but no input about that from you Microsoft, so that's the only way but is not much interesting for external contributors. We need to address project.json/xproj/dnx because is the way forward, and is good ( project.json.lock excluded )

**Comment by dsyme** [comment]
FsSrGen source is [here](https://github.com/fsharp/fsharp/tree/fsharp4/lib/bootstrap/src/FsSrGen). This has not had any changes AFAIK. There is no nuget package for this though we could create one. However this whole thing should probably be an F# script run through FSI.EXE.

FSharp.SRGen.Build.Tasks is [here](https://github.com/fsharp/fsharp/tree/fsharp4/lib/bootstrap/src/FSharp.SRGen.Build.Tasks).  This is the one that should never have existed.

As a first step it would probably be good to dump these into an https://github.com/fsprojects  project like FsLexYacc.

Stepping back, it's interesting to consider the overall cost of creating and maintaining both .NET Core and .NET Fx versions of command line tools like these (FsLexYacc.exe, FsSrGen.exe, FindUnused.exe).  There are zillions of such tools in the world.  Some questions

- I'm not exactly sure how we are meant to distribute both .NET Fx and .NET Core versions of these kinds of tools.  Currently nuget packages can have a "tools" directory and I suppose we can have multiple subdirectories there within the one nuget package.  But what do we do about targets files etc?

- Is it possible to compile a tool as portable? 

So rather than ripping out this tool maybe we should look at how we minimize costs for these kinds of tools in the .NET Core and .NET Fx world.

**Comment by enricosada** [comment]
For FsLex/FsYacc i think is great to have in project in fsprojets and should be maintaned as nuget 

FsSrGen instead is used only in this repo, in 4 locations, is more code of source/project then the resx generated
I have seen source, but need work to use nuget instead of lkg binaries, so i was asking **if is worth have it when in future we need to change it for coreclr/maintenance** but maybe i was not clear about my question 

As stand alone library outside fsharp repo, FsSrGen is easier with a type provider than msbuild task ( that doesnt work nice with nuget )

**Comment by dsyme** [comment]
@enricosada I think both type providers and msbuild tasks bring their misery as we change assumptions in the build (platform, Mono, ...).  

@KevinRansom is right - the current situation is not ideal but the cost of changing is too high (and will bring other problems)  For now we can rebuild the code easily now that I've given the link.

**Comment by enricosada** [comment]
@dsyme exactly, msbuild is annoying and type provider too much change. 
My idea is add **generated.fs and .resx** to project instead of use fssrgen. 
It's simple to do, not much diff, we can di that for each project incrementally
